staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -156,36 +174,139 @@ func apiVersionsToAPIGroup(apiVersions *metav1.APIVersions) (apiGroup metav1.API
 	return
 }
 
-// ServerGroups returns the supported groups, with information like supported versions and the
-// preferred version.
-func (d *DiscoveryClient) ServerGroups() (apiGroupList *metav1.APIGroupList, err error) {
-	// Get the groupVersions exposed at /api
-	v := &metav1.APIVersions{}
-	err = d.restClient.Get().AbsPath(d.LegacyPrefix).Do(context.TODO()).Into(v)
-	apiGroup := metav1.APIGroup{}
-	if err == nil && len(v.Versions) != 0 {
-		apiGroup = apiVersionsToAPIGroup(v)
+// GroupsAndMaybeResources returns the discovery groups, and (if new aggregated
+// discovery format) the resources keyed by group/version. Merges discovery groups
+// and resources from /api and /apis (either aggregated or not). Legacy groups
+// must be ordered first.
+func (d *DiscoveryClient) GroupsAndMaybeResources() (*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, error) {
+	// Legacy group ordered first (there is only one -- core/v1 group). Returned groups must
+	// be non-nil, but it could be empty. Returned resources, apiResources map could be nil.
+	groups, resources, err := d.downloadLegacy()
+	if err != nil {
+		return nil, nil, err
+	}
+	// Discovery groups and (possibly) resources downloaded from /apis.
+	apiGroups, apiResources, aerr := d.downloadAPIs()
+	if err != nil {
+		return nil, nil, aerr
+	}
+	// Merge apis groups into the legacy groups.
+	for _, group := range apiGroups.Groups {
+		groups.Groups = append(groups.Groups, group)
+	}
+	// If exists, merge the api resources map into the legacy resources map.
+	if resources == nil && apiResources != nil {
+		resources = map[schema.GroupVersion]*metav1.APIResourceList{}
 	}
+	if apiResources != nil {
+		for gv, resourceList := range apiResources {
+			resources[gv] = resourceList
+		}
+	}
+	return groups, resources, err
+}
+
+// downloadLegacy returns the discovery groups and possibly resources
+// for the legacy v1 GVR at /api, or an error if one occurred. It is
+// possible for the resource map to be nil if the server returned
+// the unaggregated discovery.
+func (d *DiscoveryClient) downloadLegacy() (*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, error) {
```

Comment:
`downloadLegacy` and `downloadAPIs` look very similar to me. is there a clean way to parametrize a single function to work for both?

Reply:
Yeah, I guess you could do: `downloadAggregatedOr(path string, obj any) (apidiscovery.APIGroupDiscoveryList, error)` and so either you return the discovery group list, or you fill in `obj` (which would be APIGroupList or APIVersions)?

Reply:
ah i did not see bit with `metav1.APIVersions`. that makes it trickier.

Reply:
I'd keep it like this for now.